components/raftstore/src/store/fsm/apply.rs

```diff
@@ -1652,73 +1679,145 @@ where
                         self.region
                     ));
                 }
-
-                PEER_ADMIN_CMD_COUNTER_VEC
-                    .with_label_values(&["remove_peer", "success"])
-                    .inc();
-                info!(
-                    "remove peer successfully";
-                    "region_id" => self.region_id(),
-                    "peer_id" => self.id(),
-                    "peer" => ?peer,
-                    "region" => ?&self.region
-                );
-            }
-            ConfChangeType::AddLearnerNode => {
-                PEER_ADMIN_CMD_COUNTER_VEC
-                    .with_label_values(&["add_learner", "all"])
-                    .inc();
-
-                if util::find_peer(&region, store_id).is_some() {
-                    error!(
-                        "can't add duplicated learner";
-                        "region_id" => self.region_id(),
-                        "peer_id" => self.id(),
-                        "peer" => ?peer,
-                        "region" => ?&self.region
-                    );
-                    return Err(box_err!(
-                        "can't add duplicated learner {:?} to region {:?}",
-                        peer,
-                        self.region
-                    ));
-                }
-                region.mut_peers().push(peer.clone());
-
-                PEER_ADMIN_CMD_COUNTER_VEC
-                    .with_label_values(&["add_learner", "success"])
-                    .inc();
-                info!(
-                    "add learner successfully";
-                    "region_id" => self.region_id(),
-                    "peer_id" => self.id(),
-                    "peer" => ?peer,
-                    "region" => ?&self.region,
-                );
+                (Some(exist_peer), cct) => match (exist_peer.get_role(), cct) {
+                    (PeerRole::IncomingVoter, _) | (PeerRole::DemotingVoter, _) => {
+                        error!(
```

Comment:
Perhaps should check it at the beginning to make code easier to read. And it should panic because the check has been done by raft and epoch check.